aten/src/ATen/core/type.cpp

```diff
@@ -748,6 +752,104 @@ TupleTypePtr TupleType::createNamed(
       field_types, qualName, schema)); // NOLINT(modernize-make-shared)
 }
 
+void flatten_union(TypePtr& type, std::vector<TypePtr>& res) {
+  if (type->kind() == UnionType::Kind) {
+    for (auto inner_type : type->expect<UnionType>()->types()) {
+      flatten_union(inner_type, res);
+    }
+  }
+  else if (type->kind() == OptionalType::Kind) {
+    auto inner_type = type->expect<OptionalType>()->getElementType();
+    if (inner_type->kind() == UnionType::Kind) {
+      flatten_union(inner_type, res);
+    }
+    else {
+      res.emplace_back(inner_type);
+    }
+    res.emplace_back(NoneType::get());
+  }
+  else {
+    res.emplace_back(type);
+  }
+}
+
+UnionType::UnionType(std::vector<TypePtr> types) :
+  Type(TypeKind::UnionType) {
+  // Flatten Unions of Unions while populating `types_`
+  for (auto type : types) {
+    flatten_union(type, types_);
+  }
+  // We want the elements to be sorted so we can easily compare two UnionType
+  // objects for equality in the future
+  std::sort(types_.begin(), types_.end(),
+  [](const TypePtr a, const TypePtr b) -> bool {
+    return a->kind() > b->kind();
+  });
+  // Filter out duplicate types
+  types_.erase(std::unique(types_.begin(), types_.end()), types_.end());
```

Comment:
Are you sure std::unique works here ? Won't that that be comparing pointer equality ? I know that there is a unique pointer for like Bool, but I'm not sure that guarantee exists for Lists/Dicts etc